my_enumerable.rb

```diff
@@ -1,85 +1,167 @@
+# rubocop:disable Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity, Metrics/ModuleLength
 module Enumerable
   def my_each
-    return to_enum(_method_) unless block_given?
+    return to_enum(__method__) unless block_given?
 
+    array = to_a
     i = 0
-    until i == length
-      yield(self[i])
+    until i == array.length
+      yield(array[i])
       i += 1
     end
+    if (array.is_a?(Numeric) || array.is_a?(String)) && Range.new(array.first, array.last) == self
+      self
+    else
+      array
+    end
   end
 
   def my_each_with_index
-    return to_enum(_method_) unless block_given?
+    return to_enum(__method__) unless block_given?
 
+    array = to_a
     i = 0
-    until i == length
-      yield(self[i], i)
+    until i == array.length
+      yield(array[i], i)
       i += 1
     end
+    if (array.is_a?(Numeric) || array.is_a?(String)) && Range.new(array.first, array.last) == self
+      self
+    else
+      array
+    end
   end
 
   def my_select
-    return to_enum(_method_) unless block_given?
+    return to_enum(__method__) unless block_given?
 
     array = []
     my_each { |x| array << x if yield(x) }
     array
   end
 
-  def my_all?
-    return to_enum(_method_) unless block_given?
+  def my_all?(*type)
+    origin = to_a
+    return true if origin == []
+
+    if !block_given? && type.length.zero?
+      return false if origin.include?(false) || origin.include?(nil)
+
+      return true
+    end
+
+    if type.length == 1
+      checker = origin.my_select { |x| x =~ type[0] } if type[0].instance_of?(Regexp)
+      checker = origin.my_select { |x| x == type[0] } unless type[0].instance_of?(Regexp)
 
-    array = my_select { |x| !yield(x) }
-    array.empty?
+      checker.length < origin.length ? (return false) : (return true)
+    end
+
+    checker = my_select { |x| yield(x) } if block_given? && type.length.zero?
+    checker.length < origin.length ? (return false) : (return true)
   end
 
-  def my_any?
-    return to_enum(_method_) unless block_given?
+  def my_any?(*type)
+    origin = to_a
+    return false if origin == []
+
+    if !block_given? and type.length.zero?
+      each { |x| return true if !x.nil? and x != false }
+      return false
+    end
+
+    if type.length == 1
+      checker = origin.my_select { |x| x =~ type[0] } if type[0].instance_of?(Regexp)
+      checker = origin.my_select { |x| x == type[0] } unless type[0].instance_of?(Regexp)
+      checker.empty? ? (return false) : (return true)
+    end
+
+    return true unless all? { |x| !yield(x) }
 
-    !my_all? { |x| !yield(x) }
+    false
   end
```

Comment:
- [ ]  When a class **(Numeric)** is passed as an argument, `my_any?` returns true if all of the elements of the collection is a member of that class.

```ruby
[5, 3, 3, 2].my_any?(Numeric) === [5, 3, 3, 2].any?(Numeric) => false
```